prototypes/matrix_free_stokes_global_coarsening/matrix_free_stokes_global_coarsening.cc

```diff
@@ -0,0 +1,1738 @@
+template <int dim, typename number>
+void
+StokesOperator<dim, number>::Tvmult(VectorType &      dst,
+                                          const VectorType &src) const
+{
+  this->vmult(dst, src);
+}
+
+// Computes the matrix efficiently using the optimized compute_matrix call.
+// This function is usually only used to build the system matrix for the coarse
+// grid level in the multigrid algorithm.
+template <int dim, typename number>
+const TrilinosWrappers::SparseMatrix &
+StokesOperator<dim, number>::get_system_matrix() const
+{
+  if (system_matrix.m() == 0 && system_matrix.n() == 0)
+    {
+      const auto &dof_handler = this->matrix_free.get_dof_handler();
+
+      TrilinosWrappers::SparsityPattern dsp(
+        dof_handler.locally_owned_dofs(),
+        dof_handler.get_triangulation().get_communicator());
+
+      DoFTools::make_sparsity_pattern(dof_handler, dsp, this->constraints);
+
+      dsp.compress();
+      system_matrix.reinit(dsp);
+
+      MatrixFreeTools::compute_matrix(
+        matrix_free,
+        constraints,
+        system_matrix,
+        &StokesOperator::do_cell_integral_local,
+        this);
+    }
+  return system_matrix;
+}
+
+// The diagonal of the matrix is needed, e.g., for smoothers used in
+// multigrid. Thereforem it is computed by a sequence of operator
+// evaluations to unit basis vectors using the optimized compute_diagonal
+// call.
+template <int dim, typename number>
+void
+StokesOperator<dim, number>::compute_inverse_diagonal(
+  VectorType &diagonal) const
+{
+  this->matrix_free.initialize_dof_vector(diagonal);
+  MatrixFreeTools::compute_diagonal(
+    matrix_free, diagonal, &StokesOperator::do_cell_integral_local, this);
+
+  for (auto &i : diagonal)
+    i = (std::abs(i) > 1.0e-10) ? (1.0 / i) : 1.0;
+}
+
+template <int dim, typename number>
+void
+StokesOperator<dim, number>::do_cell_integral_local(
+  FECellIntegrator &integrator) const
+{
+  using FECellIntegratorType =
+    FEEvaluation<dim, -1, 0, dim + 1, double, VectorizedArray<double>>;
+
+  FullSourceTerm<dim> source_term_function;
+
+  integrator.evaluate(EvaluationFlags::values | EvaluationFlags::gradients |
+                      EvaluationFlags::hessians);
+
+  const unsigned int cell = integrator.get_current_cell_index();
+
+  VectorizedArray<number> tau = VectorizedArray<number>(0.0);
+  std::array<number, VectorizedArray<number>::size()> h_k;
+  std::array<number, VectorizedArray<number>::size()> h;
+
+  for (auto lane = 0u; lane < matrix_free.n_active_entries_per_cell_batch(cell);
+       lane++)
+    {
+      h_k[lane] = matrix_free.get_cell_iterator(cell, lane)->measure();
+    }
+
+  for (unsigned int v = 0; v < VectorizedArray<number>::size(); ++v)
+    {
+      if (dim == 2)
+        {
+          h[v] = std::sqrt(4. * h_k[v] / M_PI);
+        }
+      else if (dim == 3)
+        {
+          h[v] = std::pow(6 * h_k[v] / M_PI, 1. / 3.);
+        }
+      tau[v] = h[v] * h[v];
+    }
```

Comment:
I guess this can be all precomputed once and read with https://www.dealii.org/developer/doxygen/deal.II/classFEEvaluationData.html#aafc3a903ca65fdbc195a17e1de9b62cf.